File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/bookmark/BookmarkDtos.kt
package dev.stackverse.backend.bookmark

import com.fasterxml.jackson.annotation.JsonInclude
import java.time.Instant
import java.util.UUID

data class BookmarkRequest(
    val url: String? = null,
    val title: String? = null,
    val notes: String? = null,
    val tags: List<String>? = null,
    val visibility: Visibility? = null,
)

@JsonInclude(JsonInclude.Include.NON_NULL)
data class BookmarkResponse(
    val id: UUID,
    val url: String,
    val title: String,
    val notes: String?,
    val tags: List<String>,
    val visibility: Visibility,
    val status: BookmarkStatus,
    val owner: String,
    val createdAt: Instant,
    val updatedAt: Instant,
) {
    companion object {
        fun of(bookmark: Bookmark) = BookmarkResponse(
            id = bookmark.id,
            url = bookmark.url,
            title = bookmark.title,
            notes = bookmark.notes,
            tags = bookmark.tags.sorted(),
            visibility = bookmark.visibility,
            status = bookmark.status,
            owner = bookmark.owner,
            createdAt = bookmark.createdAt,
            updatedAt = bookmark.updatedAt,
        )
    }
}

@JsonInclude(JsonInclude.Include.NON_NULL)
data class BookmarkCursorPageResponse(
    val items: List<BookmarkResponse>,
    val nextCursor: String?,
)

data class TagCountResponse(val tag: String, val count: Long)

data class TagListResponse(val tags: List<TagCountResponse>)

File: backends/dotnet/src/StackverseBackend/Bookmarks/BookmarkDtos.cs
namespace StackverseBackend.Bookmarks;

public sealed record BookmarkRequest(
    string? Url = null,
    string? Title = null,
    string? Notes = null,
    List<string>? Tags = null,
    Visibility? Visibility = null);

public sealed record BookmarkResponse(
    Guid Id,
    string Url,
    string Title,
    string? Notes,
    IReadOnlyList<string> Tags,
    Visibility Visibility,
    BookmarkStatus Status,
    string Owner,
    DateTime CreatedAt,
    DateTime UpdatedAt)
{
    public static BookmarkResponse Of(Bookmark bookmark) => new(
        bookmark.Id,
        bookmark.Url,
        bookmark.Title,
        bookmark.Notes,
        bookmark.Tags.Order(StringComparer.Ordinal).ToList(),
        bookmark.Visibility,
        bookmark.Status,
        bookmark.Owner,
        bookmark.CreatedAt,
        bookmark.UpdatedAt);
}

public sealed record BookmarkCursorPageResponse(IReadOnlyList<BookmarkResponse> Items, string? NextCursor);

public sealed record TagCountResponse(string Tag, long Count);

public sealed record TagListResponse(IReadOnlyList<TagCountResponse> Tags);
